### ISSUE TITLE
Task exclusion on `GradleBuild` tasks does not work past Gradle version 6.7.1

### ISSUE THREAD

**Issue by sugarmanz** [gradle/gradle#18196]
### Summary
Per the [docs](https://docs.gradle.org/current/userguide/command_line_interface.html#sec:excluding_tasks_from_the_command_line), one should be able to exclude a task from execution using the `-x` flag:
```
./gradlew someTask -x someDependentTask
```

The above would execute `someTask` while not executing `someDependentTask` even if `someTask.dependsOn(someDependentTask)`.

This still works for general tasks, but no longer works with tasks being run within a `GradleBuild` task. 

```kotlin
tasks {
    val first by creating {
        doLast {
            println("first")
        }
    }

    val second by creating {
        doLast {
            println("second")
        }

        dependsOn(first)
    }

    task<GradleBuild>("both") {
        // ensure to pass excluded tasks
        startParameter = gradle.startParameter.newInstance()

        tasks = listOf(
            first.name,
            second.name
        )

        doLast {
            println("both")
        }
    }
}
```

### Expected Behavior

On version 6.7.1:
```
./gradlew second -x first
// second

./gradlew both -x second
// first
// both
```

### Current Behavior

Post version 6.7.1:
```
./gradlew second -x first
// second

./gradlew both -x second
// first
// second
// both
```

### Context

This method of excluding tasks is heavily relied on in an open source versioning tool I help maintain for Gradle:
https://github.com/intuit/auto

The Gradle plugin for Auto uses the [ResearchGate Gradle Release Plugin](https://github.com/researchgate/gradle-release) under the hood to perform the actual version upgrade. The release plugin exposes a top-level `GradleBuild` task, `release` configured to execute the release. However, with the way Auto is structured, we need to ensure all commits are done by Auto, so we exclude certain tasks from `release`, such as the `commitNewVersion` task.

https://github.com/intuit/auto/issues/2068

### Steps to Reproduce 

I made a repo to help exemplify this issue:
https://github.com/sugarmanz/gradle-exclude-task-issue

### Your Environment

Build scan URL: https://scans.gradle.com/s/f7fk6s27kt7ua

**Comment by jbartok** [comment]
@sugarmanz Looks like the change in behaviour is actually a fix (ie. the old behaviour was incorrect). Excludes from the command-line should never have affected any tasks run by `GradleBuild`. Only the excludes specified on the `GradleBuild` task itself should apply to the tasks executed by it.

The official recommendation is to avoid using the `GradleBuild` task type entirely (see [here](https://docs.gradle.org/current/userguide/authoring_maintainable_build_scripts.html#sec:avoiding_use_of_gradlebuild)). If the alternatives suggested aren't satisfying you can always get the old behaviour by attaching the excludes from `gradle.startParameter` to the task.

**Comment by sugarmanz** [comment]
@jbartok, thanks for the information. I'm a little confused by the following tho:

> If the alternatives suggested aren't satisfying you can always get the old behaviour by attaching the excludes from gradle.startParameter to the task.

I thought I _was_ attaching the excludes from the `gradle.startParameter` in the example I posted:
```
task<GradleBuild>("both") {
    // ensure to pass excluded tasks
    startParameter = gradle.startParameter.newInstance()
    // ...
}
```

Could you provide an example of how I would correctly attach the excludes? With the code I just reference, the excluded tasks are still executed.

**Comment by jbartok** [comment]
Yes, you are right, there might be indeed a bug here. We will look into it again, as soon as we can.

**Comment by sugarmanz** [comment]
@jbartok or whoever ends up taking a look at this, I did some investigating and found that it's related to the `CompositeAwareTaskSelector` introduced in https://github.com/gradle/gradle/pull/14894. Specifically, the edge case for detecting if [`gradle.isRootBuild()`](https://github.com/gradle/gradle/blob/master/subprojects/core/src/main/java/org/gradle/execution/CompositeAwareTaskSelector.java#L58). I pushed up a commit that removes this check and it fixes this use case and doesn't seem to break any other tests, however, this seems like an obviously obtuse solution. It almost seems to me that the `CompositeAwareTaskSelector` doesn't fit this specific use case, but I don't have enough context to know for sure.

https://github.com/gradle/gradle/compare/master...sugarmanz:gradle-build-exclude-filter

Let me know if I can be of any more assistance.

**Comment by sugarmanz** [comment]
@jbartok just wanted to follow up here. This issue is causing some internal churn for our teams here at Intuit that would like to update to latest versions of Gradle but cannot due to this incompatible behavior. Please let me know if there is any plan to address this.

**Comment by jbartok** [comment]
@sugarmanz yes, we definitely plan to address this, we're just lacking the necessary capacity for it. Let me see if I can speed up the process.

**Comment by ykhandelwal913** [comment]
@jbartok just wanted to follow up here.  Our teams would like to update to latest gradle version using auto but cannot due to this behaviour. Please let me know if there is any plan to address this.

**Comment by jbartok** [comment]
@ykhandelwal913 sorry for the inconvenience. I've attempted a fix (in the attached PR) but turns out I've bitten off more than I can chew... So now I'm waiting for help from more senior developers, which is not really forthcoming. I will attempt to push, but I'm not optimistic...

**Comment by nigelsim** [comment]
This is the current work around we are using with the release plugin. It is probably incomplete, but works well enough for us.

```groovy
project.gradle.taskGraph.whenReady {
    project.gradle.taskGraph.getAllTasks().forEach { task -> if (project.gradle.startParameter.excludedTaskNames.contains(task.name)) { task.setEnabled(false) } }
}
```

**Comment by jbartok** [comment]
@sugarmanz, @ykhandelwal913, @nigelsim: I'm happy to say that (through no merit of my own), the issue seems to be fixed from Gradle version `7.6` onward.
